File: javastory.budgetsh.stage4/src/javastory/budgetsh/stage4/server/budget/da/file/io/CashBookFile.java
package javastory.budgetsh.stage4.server.budget.da.file.io;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.Writer;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.Map;

import javastory.budgetsh.stage4.server.budget.entity.budget.CashBook;
import javastory.budgetsh.stage4.server.club.da.file.io.FileConfig;
import javastory.budgetsh.stage4.server.club.util.FileDbWrapper;

public class CashBookFile {
	//
	private FileDbWrapper cashbookFile;
	private FileDbWrapper cashbookTempFile;
	
	private static String defaultDelimeter = "/";
	
	private static Map<String, Integer> keyIndexMap; 
	
	static {
		keyIndexMap = new LinkedHashMap<String, Integer>();
		keyIndexMap.put("bankAccount", 0);
	}
	
	public CashBookFile() {
		//
		cashbookFile = new FileDbWrapper(
				"Budget",
				FileConfig.getFileName("CashBook"),
				FileConfig.getDelimiter());
		cashbookTempFile = new FileDbWrapper(
				"Budget",
				FileConfig.getFileName("CashBookTemp"),
				FileConfig.getDelimiter());
		
		cashbookFile.setKeyIndexMap(keyIndexMap);
		cashbookTempFile.setKeyIndexMap(keyIndexMap);
	}
	
	public boolean exist(String bankAccount) {
		// 
		boolean found = false;
		BufferedReader reader;

		String line;
		try {
			reader = cashbookFile.requestReader();
			while(!((line = reader.readLine()) == null)) {
				if(cashbookFile.hasValueOf("bankAccount",bankAccount,line)) {
					found =true;
					break;
				}
			}
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		return found;
	}
	
	public CashBook retrieve(String bankAccount) {
		//
		CashBook foundCashBook = null;
		
		BufferedReader reader;
		String line;
		try {
			reader = cashbookFile.requestReader();
			while(!((line = reader.readLine()) == null)) {
				if(cashbookFile.hasValueOf("bankAccount",bankAccount, line)) {
					foundCashBook = this.convertToCashBook(line);
					break;
				}
			}
			reader.close();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		
		return foundCashBook;
	}

	public void remove(CashBook cashbook) {
		//
		String account = cashbook.getBankAccount();
		BufferedReader reader;
		PrintWriter writer;
		String line;
		
		try {
			reader = cashbookFile.requestReader();
			writer = cashbookTempFile.requestPrintWriter();
			while(!((line = reader.readLine()) == null)) {
				if(cashbookFile.hasValueOf("bankAccount",account,line)) {
					continue;
				}
				writer.write(line);
				writer.write("\r\n");
				
				writer.close();
				reader.close();
			}
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		
		if(!cashbookFile.delete()) {
			System.out.println("failure to delete");
		}
		
		if(!cashbookTempFile.renameTo(cashbookFile)) {
			System.out.println("Failure to rename");
		}
	}

	public void update(CashBook cashbook) {
		//
		String bankAccount = cashbook.getBankAccount();
		BufferedReader reader;
		PrintWriter writer;
		
		try {
			reader = cashbookFile.requestReader();
			writer = cashbookTempFile.requestPrintWriter();
			
			String line;
			
			while(!((line = reader.readLine()) == null)) {
				if((cashbookFile.hasValueOf("bankAccount",bankAccount,line))) {
					writer.write(cashbookTempFile.convertFrom(cashbook));
					writer.write("\r\n");
				}else {
					writer.write(line);
					writer.write("\r\n");
				}
			}
			writer.close();
			reader.close();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		
		if(!cashbookFile.delete()) {
			System.out.println("failure to delete");
			return;
		}
		
		if(!cashbookTempFile.renameTo(cashbookFile)) {
			System.out.println("Failure to rename");
		}
	}

	public void regist(CashBook cashbook) {
		//
		Writer writer;
		try {
			writer = cashbookFile.requestPrintWriter();
			writer.write(cashbookFile.convertFrom(cashbook));
			writer.write("\r\n");
			writer.close();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}
	
	private CashBook convertToCashBook(String line) {
		//
		return (CashBook) cashbookFile.convertTo(line, CashBook.class);
	}
	
	private String convertToLine(CashBook cashbook) {
		//
		return cashbookFile.convertFrom(cashbook);
	}

	public ArrayList<CashBook> retrieveAll() {
		//
		CashBook foundCashBook = null;
		ArrayList<CashBook> cashbookList = new ArrayList<CashBook>();
		
		BufferedReader reader;
		String line;
		try {
			reader = cashbookFile.requestReader();
			while(!((line = reader.readLine()) == null)) {
					foundCashBook = this.convertToCashBook(line);
					cashbookList.add(foundCashBook);
			}
			reader.close();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		
		return cashbookList;
	}


}

File: javastory.budgetsh.stage5.1/src/step3/server/store/filestore/io/CashbookFile.js
const FileDbWrapper = require('./FileDbWrapper.js');
const defaultDelimiter = '||';

const cashbookFile = new FileDbWrapper('Resource','cashbook',defaultDelimiter);
const cashbookTempFile = new FileDbWrapper('Resource', 'cashbookTemp', defaultDelimiter);
const keyIndexMap = new Map();
keyIndexMap.set('bankAccount', 0);

cashbookFile.keyIndexMap = keyIndexMap;
cashbookTempFile.keyIndexMap = keyIndexMap;

exports.exist = function(bankAccount){
	let found = false;

	let lines = [];
	lines = cashbookFile.readFile();
	
	let lineSize = lines.length;
	for(var i = 0; i<lineSize; i++){
		if(cashbookFile.hasValueOf('bankAccount', bankAccount, lines[i])){
			found = true;
			break;
		}
	}
	return found;
}

exports.regist = function(cashbook){
	//
	let data = JSON.stringify(cashbook);
	cashbookFile.writeFile(cashbook.bankAccount);
	cashbookFile.writeFile(defaultDelimiter);
	cashbookFile.writeFile(data);
	cashbookFile.writeFile('\r\n');
}

exports.retrieve = function(bankAccount){
	let cashbook = null;
	let lines = [];
	let json = '';
	lines = cashbookFile.readFile();
	
	let lineSize = lines.length;
	for(var i = 0; i<lineSize; i++){
		if(cashbookFile.hasValueOf('bankAccount', bankAccount, lines[i])){
			json = lines[i].split(defaultDelimiter)[1];
			cashbook = JSON.parse(json);
			break;
		}
	}
	
	return cashbook;
}

exports.update = function(cashbook){
	let account = cashbook.bankAccount;
	let lines = [];
	let json = '';
	lines = cashbookFile.readFile();
	let lineSize = lines.length;
	for(var i = 0; i<lineSize; i++){
		if(cashbookFile.hasValueOf('bankAccount', account, lines[i])){
			json = JSON.stringify(cashbook);	
			cashbookTempFile.writeFile(cashbook.bankAccount);
			cashbookTempFile.writeFile(defaultDelimiter);
			cashbookTempFile.writeFile(json);
			cashbookTempFile.writeFile('\r\n');
		}else{
			cashbookTempFile.writeFile(lines[i]);
			cashbookTempFile.writeFile('\r\n');
		}
	}
	cashbookTempFile.writeFile('\r\n');
	cashbookTempFile.deleteFile(cashbookFile.getFilePath());
	cashbookTempFile.rename(cashbookFile.getFilePath());
}

exports.remove = function(cashbook){
	let foundCashbook = null;
	let cashbookAccount = cashbook.bankAccount;
	let lines = [];
	let json = '';
	lines = cashbookFile.readFile();
	let lineSize = lines.length;
	for(var i = 0; i<lineSize; i++){
		if(!cashbookFile.hasValueOf('bankAccount', cashbookAccount, lines[i])){
			cashbookTempFile.writeFile(lines[i]);
			cashbookTempFile.writeFile('\r\n');
		}
	}
	cashbookTempFile.writeFile('\r\n');
	cashbookTempFile.deleteFile(cashbookFile.getFilePath());
	cashbookTempFile.rename(cashbookFile.getFilePath());
}